p2p/src/net/mock/types.rs

```diff
@@ -199,13 +205,20 @@ pub enum MockEvent {
     SendMessage(Box<Message>),
 }
 
+// TODO: Decide what to do about protocol upgrades.
+// For example adding new address type to PeerAddress might break handshakes with older nodes.
 #[derive(Debug, Encode, Decode, Clone, PartialEq, Eq)]
 pub enum HandshakeMessage {
     Hello {
         version: common::primitives::semver::SemVer,
         network: [u8; 4],
         protocols: Vec<Protocol>,
         subscriptions: BTreeSet<PubSubTopic>,
+
+        // Specify on which port this node can be reached for incoming connections.
+        // If not set, then this node does not want incoming connections.
+        // Actual IP is taken from the remote socket address.
+        listening_port: Option<u16>,
```

Comment:
This is a weird thing... why? When someone connects, they advertise their socket address through their connection... I kind of don't understand this.

Reply:
When you connect with TCP OS chooses some random port for you. That is because listening socket and connecting sockets are independent from the OS viewpoint.

Reply:
Still. That isn't really the way this should work. The way I see it (and correct me if I'm wrong here), a node connects to another node, and broadcasts the addresses it finds working, including its own. Ports should never be expected to be different.

This even ignores NAT issues. Behind NAT subnets, no one can know this value because the router decides it and even the value the node thinks it has is firewalled, which basically means our software should be able to detect whether a node is in a private network... why bother?

How is this even gonna be verified? Are we gonna create a second hand-shake process to ensure that the sender isn't lying or something?

I think this should go, and we should just expect ports to be part of the broadcast addresses at worst.

Reply:
Sure, different ways are possible.
The main question is how the node will find own IP address.
1. The node might try detect own IP addresses using OS specific API (I think that is what you propose).
2. Remote peers can read listening port for all incoming connections and get peer address directly (what is implemented here).

Problems with the first approach:
- Need to use OS specific API, IP addresses might change with time so that should be done periodically.
- Does not work with NAT and port forwarding (when forwarded port same).

The second approach does not have this cons and is easier and so I decided to use it.
Grin works this way. In bitcoin core public IP also loaded from remote peers, but handled locally instead.

> This even ignores NAT issues.

I don't see a problem here. And this is done specifically to handle NAT with port forwarding, because it's difficult to find out public IP address.

> which basically means our software should be able to detect whether a node is in a private network... why bother?

We don't need to detect if we are in a private network. Remote peer will only see public IP address.

> How is this even gonna be verified? Are we gonna create a second hand-shake process to ensure that the sender isn't lying or something?

This new address will be verified same way like any other discovered address. The node might not be reachable but that is not a problem.

Reply:
Maybe there's a misunderstanding, but I have mentioned many times that we're not gonna do anything to circumvent NAT. No hole-punching. Nothing. If someone is interested in using DMZ to open their network to the mintlayer network, they should do it. We shouldn't attempt to do anything for NAT.

You're saying "We don't need to detect if we are in a private network. Remote peer will only see public IP address"... it might be a misunderstanding also on how NAT works. Even if we see a public IP address, this doesn't mean that this IP address will work, because again, it's behind a firewall and a NAT mapping should be created beforehand to make that work. So the question is, if someone is going to configure their router to do that mapping through DMZ or some other routing/mapping, what use does this port value serve exactly? We aren't expecting people to use arbitrary port numbers. Bitcoin doesn't do that either.

And just to give you an example from real life, apart from bitcoin which does exactly that (please correct me if I'm wrong here), the SMTP protocol for MTA (Mail Transfer Agents) works on port 25. No one expects other agents to use other ports. No one broadcasts different ports. No one tries to circumvent NAT to reach these ports. It's the administrator's job to maintain that promise. Same for http, https, etc.

You're saying we need OS-specific APIs. How do other blockchains deal with this? How does substrate deal with this? Grin? There are many other blockchains in rust. Check them out, please.

Reply:
One point I wanna add here is that the best we can do here is make peers broadcast socket addresses instead of just addresses. This way people can announce hosts with different ports, but it's not the default behavior. I someone wants to break that promise, they have to go through the work of broadcasting that address. Now the automation of this process is a whole other question. Again, let's see how other blockchains do this.

Reply:
I looked at Bitcoin, Monero and Grin (I skipped Bitcoin forks and blockchains that use libp2p).
In all three, public node addresses might be constructed using socket addresses, as seen by remote peers and local listening ports.

Bitcoin:

```
if (!pfrom.IsInboundConn() && SetupAddressRelay(pfrom, *peer)) {
    if (fListen && !m_chainman.ActiveChainstate().IsInitialBlockDownload())
    {
        CAddress addr{GetLocalAddress(pfrom.addr), peer->m_our_services, Now<NodeSeconds>()};
        if (addr.IsRoutable())
        {
            LogPrint(BCLog::NET, "ProcessMessages: advertising address %s\n", addr.ToString());
            PushAddress(*peer, addr, insecure_rand);
        } else if (IsPeerAddrLocalGood(&pfrom)) {
            addr.SetIP(addrMe);
            LogPrint(BCLog::NET, "ProcessMessages: advertising address %s\n", addr.ToString());
            PushAddress(*peer, addr, insecure_rand);
        }
    }
}
```

So if a node behind NAT and made an outgoing connection, `GetLocalAddress` will return local listining address and the node will change IP address to public socket address and use that to advertise as its own address.

Grin:

```
/// Resolve the correct peer_addr based on the connection and the advertised port.
fn resolve_peer_addr(advertised: PeerAddr, conn: &TcpStream) -> PeerAddr {
    let port = advertised.0.port();
    if let Ok(addr) = conn.peer_addr() {
        PeerAddr(SocketAddr::new(addr.ip(), port))
    } else {
        advertised
    }
}

pub fn accept(
    &self,
    capab: Capabilities,
    total_difficulty: Difficulty,
    conn: &mut TcpStream,
) -> Result<PeerInfo, Error> {
...
    // all good, keep peer info
    let peer_info = PeerInfo {
        ...
        addr: resolve_peer_addr(hand.sender_addr, &conn),
    };
```

When a new peer is connected, IP address of the socket and advertised port is used to construct a public address of this peer.

Monero:

```
  template<class t_payload_net_handler>
  int node_server<t_payload_net_handler>::handle_handshake(int command, typename COMMAND_HANDSHAKE::request& arg, typename COMMAND_HANDSHAKE::response& rsp, p2p_connection_context& context)
  {
    if(!context.m_is_income)
    {
      return 1;
    }

    if(arg.node_data.my_port && zone.m_can_pingback)
    {
      uint32_t port_l = arg.node_data.my_port;
      //try ping to be sure that we can add this peer to peer_list
      try_ping(arg.node_data, context, [peer_id_l, port_l, context, this]()
      {
        //called only(!) if success pinged, update local peerlist
...
  }

  bool node_server<t_payload_net_handler>::get_local_node_data(basic_node_data& node_data, const network_zone& zone)
  {
    if(!m_hide_my_port && zone.m_can_pingback)
      node_data.my_port = m_external_port ? m_external_port : m_listening_port;
    else
      node_data.my_port = 0;
  }

```

For incoming connections `try_ping` is called, to check if a remote node is also accepting incoming connections. `my_port` is set by the remote peer itself and IP is taken from the remote socket address.

So in all three blockchains the same mechanism is used, in one way or another:
When a node is accepting an incoming connection, it takes IP address from remote socket and advertised listening port. This new socket address is then used as a new potential address.
This approach is universal enough to work with different setups, there is no need to load interface addresses manually (although bitcoin loads them).
And that is what I have implemented here.

BTW, I found two popular crates that can be used to find local network interface addresses (`if-addrs` and `if-watch`). I missed them before.